### PR TITLE
AAC-591 - Add Struct log

### PR DESCRIPTION
**Pull request by pking-moj**
## Description of change

Move to using structured logging for better searching when in Kibana

## Link to Jira Ticket

- [AAC-591](https://dsdmoj.atlassian.net/browse/AAC-591)

## Screenshots or test evidence if applicable

Screenshot 2022-05-04 at 21 01 28

### CODE REVIEW

File: laa_court_data_api_app/routers/laa_references.py [hunk 1]
```diff
@@ -38,7 +38,7 @@ async def patch_maat_unlink(defendant_id: str, request: ExternalPatchRequest):
         return JSONResponse(status_code=400, content=LaaReferencesErrorResponse(
             error={'defendant_id': ['mismatch in ids given']}).dict())
 
-    logger.info(f"Calling_Maat_Patch_{defendant_id}")
+    logger.info("Calling_Maat_Patch", defendant_id=defendant_id)
```
Comment (wordshaker) [inline laa_court_data_api_app/routers/laa_references.py:41]: I know it's the convention you are using, and don't change it, BUT seeing MAAT as Maat is making my retinas itch

File: laa_court_data_api_app/routers/laa_references.py [hunk 2]
```diff
@@ -48,39 +48,39 @@ async def patch_maat_unlink(defendant_id: str, request: ExternalPatchRequest):
...
+    cda_response = await client.post("/api/internal/v2/laa_references/",
                                      body=InternalPostRequest(laa_reference=InternalPost(**request.dict())))
 
     return formulated_response(cda_response, request.defendant_id, "Linking")
 
 
 def formulated_response(cda_response, defendant_id, request_type):
     if cda_response is None:
-        logging.error("Laa_References_Endpoint_Did_Not_Return")
+        logger.error("Laa_References_Endpoint_Did_Not_Return")
```
Comment (wordshaker) [inline laa_court_data_api_app/routers/laa_references.py:62]: I'm guessing the defendant_id wouldn't help with debugging and thats why it's not included in the error log?

**Review by wordshaker — APPROVED** [review]

File: laa_court_data_api_app/routers/laa_references.py (hunk 2 above)
Comment (pking-moj) [reply laa_court_data_api_app/routers/laa_references.py]: Every little helps! added info

File: laa_court_data_api_app/routers/laa_references.py (hunk 1 above)
Comment (pking-moj) [reply laa_court_data_api_app/routers/laa_references.py:41]: 😿

File: laa_court_data_api_app/routers/hearing_events.py [hunk 3]
```diff
@@ -8,32 +8,32 @@
...
 
     match cda_response.status_code:
         case 200:
-            logging.info("Hearing_Events_Endpoint_Returned_Success")
+            logger.info("Hearing_Events_Endpoint_Returned_Success")
             hearing_events_result = HearingEventsResult(**cda_response.json())
             return HearingEventsResponse(**hearing_events_result.dict())
         case 400:
-            logging.info("Hearing_Events_Endpoint_Validation_Failed")
+            logger.info("Hearing_Events_Endpoint_Validation_Failed")
```
Comment (wordshaker) [inline laa_court_data_api_app/routers/hearing_events.py:32]: Again info not `warn`? and do we want to add status code to the log?

File: laa_court_data_api_app/routers/hearing_events.py [hunk 4]
```diff
@@ -8,32 +8,32 @@
...
             return Response(status_code=400)
         case 404:
-            logging.info("Hearing_Events_Endpoint_Not_Found")
+            logger.info("Hearing_Events_Endpoint_Not_Found")
             return Response(status_code=404)
         case _:
-            logging.error("Hearing_Events_Endpoint_Error_Returning")
+            logger.error("Hearing_Events_Endpoint_Error_Returning")
```
Comment (wordshaker) [inline laa_court_data_api_app/routers/hearing_events.py:38]: Again info not `warn`? and do we want to add status code to the log?

File: laa_court_data_api_app/routers/hearing_summaries.py [hunk 5]
```diff
@@ -11,38 +11,38 @@
...
-            logging.info("Prosecution_Case_Endpoint_Validation_Failed")
+            logger.info("Prosecution_Case_Endpoint_Validation_Failed")
             return Response(status_code=400)
         case 404:
-            logging.info("Prosecution_Case_Endpoint_Not_Found")
+            logger.info("Prosecution_Case_Endpoint_Not_Found")
             return Response(status_code=404)
         case _:
-            logging.error("Prosecution_Case_Endpoint_Error_Returning")
+            logger.error("Prosecution_Case_Endpoint_Error_Returning")
```
Comment (wordshaker) [inline laa_court_data_api_app/routers/hearing_summaries.py:45]: Could add status code to help with tracing through logs on incident

File: laa_court_data_api_app/routers/hearing_summaries.py [hunk 6]
```diff
@@ -11,38 +11,38 @@
...
+            logger.info("Prosecution_Case_Endpoint_Returned_Success")
             prosecution_case_results = ProsecutionCasesResults(**cda_response.json())
             summaries = map_hearing_summaries(prosecution_case_results.results)
-            logging.info(f"Hearing_Summaries_To_Show: {summaries.count}")
+            logger.info("Hearing_Summaries_To_Show", count=len(summaries))
             return HearingSummariesResponse(hearing_summaries=summaries,
                                             overall_defendants=map_defendant_list(prosecution_case_results.results))
         case 400:
-            logging.info("Prosecution_Case_Endpoint_Validation_Failed")
+            logger.info("Prosecution_Case_Endpoint_Validation_Failed")
```
Comment (wordshaker) [inline laa_court_data_api_app/routers/hearing_summaries.py:39]: Could add status code to help with tracing through logs on incident

Could also be a `warn` for a bad request

File: laa_court_data_api_app/routers/hearing_summaries.py [hunk 7]
```diff
@@ -11,38 +11,38 @@
...
     logger.info("Calling_Hearing_Summaries_Get_Endpoint")
-    logging.info(f"Hearing_Summaries_Get_{urn}")
+    logger.info("Hearing_Summaries_Get", urn=urn)
     client = CourtDataAdaptorClient()
     cda_response = await client.get("/api/internal/v2/prosecution_cases",
                                     params={"filter[prosecution_case_reference]": urn})
 
     if cda_response is None:
-        logging.error("Prosecution_Case_Endpoint_Did_Not_Return")
+        logger.error("Prosecution_Case_Endpoint_Did_Not_Return")
```
[the remark this answers is not in the document]
Comment (pking-moj) [reply laa_court_data_api_app/routers/hearing_summaries.py:26]: If the response is none then nothing go returned. No status code or anything

File: laa_court_data_api_app/routers/laa_references.py [hunk 8]
```diff
@@ -48,39 +48,39 @@ async def patch_maat_unlink(defendant_id: str, request: ExternalPatchRequest):
...
             return JSONResponse(status_code=400,
                                 content=LaaReferencesErrorResponse(
                                     error=cda_response.json()).dict())
         case 404:
-            logging.info(f"Laa_References_Endpoint_Not_Found")
+            logger.info(f"Laa_References_Not_Found_For_{request_type}", defendant_id=defendant_id)
             return Response(status_code=404)
         case 422:
-            logging.info(f"Unable_To_Process_{request_type}_For_{defendant_id}")
+            logger.info(f"Unable_To_Process_{request_type}", defendant_id=defendant_id)
```
Comment (wordshaker) [inline laa_court_data_api_app/routers/laa_references.py:78]: Do we want status code in here to help with tracing issues?

File: laa_court_data_api_app/routers/laa_references.py [hunk 9]
```diff
@@ -48,39 +48,39 @@ async def patch_maat_unlink(defendant_id: str, request: ExternalPatchRequest):
...
         return Response(status_code=424)
 
     match cda_response.status_code:
         case 202:
-            logging.info(f"Maat_Id_For_{defendant_id}_{request_type}_Successfully_Requested")
+            logger.info(f"MAAT_Id_For_{request_type}_Successfully_Requested", defendant_id=defendant_id)
             return Response(status_code=202)
         case 400:
-            logging.info(f"Validation_Failed_For_{defendant_id}")
+            logger.info(f"Validation_Failed_For_{request_type}", defendant_id=defendant_id)
```
Comment (wordshaker) [inline laa_court_data_api_app/routers/laa_references.py:70]: Do we want Bad Requests as a warn?

File: laa_court_data_api_app/routers/defendants.py [hunk 10]
```diff
@@ -25,51 +25,52 @@ async def get_defendants(urn: str | None = None,
...
         cda_response = await client.get("/api/internal/v2/prosecution_cases",
                                         params={"filter[national_insurance_number]": nino})
     else:
         logger.error("Invalid_Defendant_Search")
         return Response(status_code=400)
 
     if cda_response is None:
-        logging.error("Prosecution_Case_Endpoint_Did_Not_Return")
+        logger.error("Prosecution_Case_Endpoint_Did_Not_Return",
+                     urn=urn, name=name, dob=dob, uuid=uuid, asn=asn, nino=nino)
```
[the remark this answers is not in the document]
Comment (wordshaker) [reply laa_court_data_api_app/routers/defendants.py:52]: We should be logging PII and you've got a uuid for tracing. Personally I'd strongly advise not logging PII for the reasons you know.

**Review by OliverBriggs — APPROVED** [review]

**Review by Obsiye — APPROVED** [review]

**Review by ivanELEC — APPROVED** [review]

**Pull request merged**